### PR TITLE
Correctly handle trailing commas that are inside a line's leading non-nested parens

### LINKED ISSUE #1671: Only the last trailing comma in a statement is magic
I would expect this to be left alone by the magic trailing comma rule:

```python
zero(
    one,
).two(
    three,
).four(
    five,
)
```

But Black implodes it to

```python
zero(one,).two(three,).four(
    five,
)
```

**Environment (please complete the following information):**

- Version: 20.8b1
- OS and Python version: Linux/Python 3.8.5

**Does this bug also happen on master?** Yes ([playground](https://black.now.sh/?version=master&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ACHAFddAD2IimZxl1N_Wm7g9A2LVPAKKhJ31iwPpbnQHIcmB_RXSDIyQOMOvlJF3j6eBFRCkSY8YOoNAXDvTLGdfxHPeWFc8DSFAE8je208jm9h59x9vFBoUgh4EAAA35vyazJE-aAAAXOIAQAAAIQ3yXexxGf7AgAAAAAEWVo=))

### LINKED ISSUE #3229: Trailing commas not removed when dict/function call collapsed into single line
**Describe the bug**

When a multiline dict/function call is collapsed into fewer lines by black, it seems to leave stray trailing commas. This does not happen when using `--skip-magic-trailing-comma`.

**To Reproduce**

For example, take this code:

```python
def refresh_token(self, device_family, refresh_token, api_key):
    return self.orchestration.refresh_token(
        data={
            "refreshToken": refresh_token,
        },
        api_key=api_key,
    )["extensions"]["sdk"]["token"]
```

And run it with these arguments:

```sh
$ black file.py -l 120
```

The resulting output is:

```python
def refresh_token(self, device_family, refresh_token, api_key):
    return self.orchestration.refresh_token(data={"refreshToken": refresh_token,}, api_key=api_key,)["extensions"][
        "sdk"
    ]["token"]
```

**Expected behavior**

The output should look something like this (without the stray trailing commas):

```python
def refresh_token(self, device_family, refresh_token, api_key):
    return self.orchestration.refresh_token(data={"refreshToken": refresh_token}, api_key=api_key)["extensions"][
        "sdk"
    ]["token"]
```

**Environment**

- Black's version: 22.6.0 / main
- OS and Python version: Debian 10, Python 3.10.4

**Additional context**

### PR DESCRIPTION
**Pull request by yilei**
### Description

Correctly handle trailing commas that are inside a line's leading non-nested parens.

- Fixes #1671
- Fixes #3229

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Comment by yilei** [comment]
Alright, this PR is ready for review. It should unblock #3368.

My initial implementation accidentally fixed the trailing comma issue in #2052, but it had other issues and was wrong so I removed that reference. I'll take another look once this PR is merged.

File: src/black/linegen.py
Comment (JelleZijlstra) [inline src/black/linegen.py:758]:
```suggestion
    if component is _BracketSplitComponent.body:
```
Nit: use `is` for comparing enums.

File: src/black/linegen.py
Comment (JelleZijlstra) [inline src/black/linegen.py:799]:
```suggestion
        and component is _BracketSplitComponent.head
```

File: src/black/linegen.py
Comment (JelleZijlstra) [inline src/black/linegen.py:811]:
```suggestion
    if component is _BracketSplitComponent.body and should_split_line(
```

**Review by JelleZijlstra — COMMENTED** [review]
Looks good!

File: src/black/linegen.py
Comment (yilei) [reply src/black/linegen.py]: Oops, removed.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**
